### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by loganwright**
#92 

Commits got a little mixed up either way, bringing us back here.

### CODE REVIEW

**Comment by andreasley** [comment]
@LoganWright I wouldn't mind having permission to create branches in `socks`. Makes merges less tricky than using my own fork.

**Comment by andreasley** [comment]
@LoganWright I'm worried about the quality of `Socks`. There are a few fundamental problems and I just keep piling on fixes, hoping not to create new problems for its users. Is switching `Vapor` to a different library worth considering (before Swift Server APIs are ready)?

**Comment by andreasley** [comment]
I have some additional fixes in https://github.com/andreasley/socks . Especially some improvements to to sending data on nonblocking sockets (used to fail for big amounts of data).

File: Sources/SocksCore/Socket.swift [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 public protocol RawSocket {
     var descriptor: Descriptor { get }
+    var closed:Bool { get }
```
Comment (tannernelson) [inline Sources/SocksCore/Socket.swift]: Minor spacing issue here.

File: Sources/SocksCore/TCPSocket.swift [hunk 2]
```diff
@@ -54,7 +54,7 @@ extension TCPReadableSocket {
         }
         
         let finalBytes = data.characters[0..<receivedBytes]
-        let out = Array(finalBytes.map({ UInt8($0) }))
+        let out = Array(finalBytes)
```
Comment (tannernelson) [inline Sources/SocksCore/TCPSocket.swift]: This will speed things up!

File: Sources/SocksCore/TCPSocket.swift [hunk 3]
```diff
@@ -104,6 +104,10 @@ public class TCPInternetSocket: InternetSocket, TCPSocket, TCPReadableSocket, TC
 
         try setReuseAddress(true)
     }
+    
+    deinit {
+        try? self.close()
```
Comment (tannernelson) [inline Sources/SocksCore/TCPSocket.swift]: Do we run any risk of exhausting open file handles if things are not closed properly and we are not warned? Perhaps we should log an error here.

File: Tests/LinuxMain.swift [hunk 4]
```diff
@@ -7,5 +7,7 @@ XCTMain([
 	testCase(LiveTests.allTests),
 	testCase(PipeTests.allTests),
 	testCase(SelectTests.allTests),
-	testCase(TimeoutTests.allTests)
+	testCase(TimeoutTests.allTests),
+	testCase(WatchingTests.allTests),
+    testCase(LifetimeTests.allTests)
```
Comment (tannernelson) [inline Tests/LinuxMain.swift]: Minor formatting issue.

File: Tests/SocksCoreTests/LifetimeTests.swift [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        
+        // release socket
+        optionalSocket = nil
+        
+        // file descriptor should be closed
+        let fdFlagClosed = fcntl(descriptor, F_GETFD)
+        XCTAssertEqual(fdFlagClosed, -1)
+    }
+}
```
Comment (tannernelson) [inline Tests/SocksCoreTests/LifetimeTests.swift]: Great tests 👍

File: Sources/SocksCore/TCPSocket.swift (hunk 3 above)
Comment (andreasley) [inline Sources/SocksCore/TCPSocket.swift]: We can't do much here; we don't even have an error handler. :(
`close()` can only fail for these reasons:
- EBADF: Invalid file descriptor => if this happens, we won't be able to do anything about it
- EINTR: Interrupter by a signal => shouldn't happen unless somebody was using the descriptor outside of `Socks`
- EIO: An I/O error occurred => We can't fix that anyway

**Comment by tannernelson** [comment]
Closing in favor of #101.

**Pull request closed without merging**